### PR TITLE
feat(ppf): Enable running in either multithread or multiprocess mode

### PR DESCRIPTION
**Pull request by nikhars**
In order to gain more performance from the post process forwarder, make it possible to run the post process forwarder in multiprocess mode. Introduced click options to enable running it it multiprocess mode and pass in the multiprocess arguments. Based on the mode, the appropriate strategy factory is instantiated.

### Testing
Added some logging for local testing and ran the ingestion pipeline locally. Note the log message for multithread mode and multiprocess mode are different.

```bash


21:36:01 system                                | post-process-forwarder-errors started (pid=45747)
21:36:18 post-process-forwarder-errors         | 21:36:18 [INFO] arroyo.processing.processor: New partitions assigned: {Partition(topic=Topic(name='events'), index=0): 147}
21:36:18 post-process-forwarder-errors         | 21:36:18 [INFO] sentry.post_process_forwarder.post_process_forwarder: Starting multithreaded post process forwarder
21:36:34 post-process-forwarder-errors         | 21:36:34 [INFO] sentry.eventstream.kafka.dispatch: Received message: Message({Partition(topic=Topic(name='events'), index=0): 148})
21:36:34 post-process-forwarder-errors         | 21:36:34 [INFO] sentry.eventstream.kafka.dispatch: Received message: Message({Partition(topic=Topic(name='events'), index=0): 149})
21:36:34 post-process-forwarder-errors         | 21:36:34 [INFO] sentry.eventstream.kafka.dispatch: Received message: Message({Partition(topic=Topic(name='events'), index=0): 150})
21:36:34 post-process-forwarder-errors         | 21:36:34 [INFO] sentry.eventstream.kafka.dispatch: Received message: Message({Partition(topic=Topic(name='events'), index=0): 151})
21:36:34 post-process-forwarder-errors         | 21:36:34 [INFO] sentry.eventstream.kafka.dispatch: Received message: Message({Partition(topic=Topic(name='events'), index=0): 152})
21:36:34 post-process-forwarder-errors         | 21:36:34 [INFO] sentry.eventstream.kafka.dispatch: Received message: Message({Partition(topic=Topic(name='events'), index=0): 153})
21:36:34 post-process-forwarder-errors         | 21:36:34 [INFO] sentry.eventstream.kafka.dispatch: Received message: Message({Partition(topic=Topic(name='events'), index=0): 154})
21:36:36 post-process-forwarder-errors         | 21:36:36 [INFO] sentry.eventstream.kafka.dispatch: Received message: Message({Partition(topic=Topic(name='events'), index=0): 155})
21:36:40 post-process-forwarder-errors         | 21:36:40 [INFO] sentry.eventstream.kafka.dispatch: Received message: Message({Partition(topic=Topic(name='events'), index=0): 156})
21:36:40 post-process-forwarder-errors         | 21:36:40 [INFO] sentry.eventstream.kafka.dispatch: Received message: Message({Partition(topic=Topic(name='events'), index=0): 157})
21:36:42 post-process-forwarder-errors         | 21:36:42 [INFO] sentry.eventstream.kafka.dispatch: Received message: Message({Partition(topic=Topic(name='events'), index=0): 158})
21:36:42 post-process-forwarder-errors         | 21:36:42 [INFO] sentry.eventstream.kafka.dispatch: Received message: Message({Partition(topic=Topic(name='events'), index=0): 159})
21:36:42 post-process-forwarder-errors         | 21:36:42 [INFO] sentry.eventstream.kafka.dispatch: Received message: Message({Partition(topic=Topic(name='events'), index=0): 160})
21:36:42 post-process-forwarder-errors         | 21:36:42 [INFO] sentry.eventstream.kafka.dispatch: Received message: Message({Partition(topic=Topic(name='events'), index=0): 161})
21:36:45 post-process-forwarder-errors         | 21:36:45 [INFO] sentry.eventstream.kafka.dispatch: Received message: Message({Partition(topic=Topic(name='events'), index=0): 162})
21:36:45 post-process-forwarder-errors         | 21:36:45 [INFO] sentry.eventstream.kafka.dispatch: Received message: Message({Partition(topic=Topic(name='events'), index=0): 163})
21:36:45 post-process-forwarder-errors         | 21:36:45 [INFO] sentry.eventstream.kafka.dispatch: Received message: Message({Partition(topic=Topic(name='events'), index=0): 164})
21:36:47 post-process-forwarder-errors         | 21:36:47 [INFO] sentry.eventstream.kafka.dispatch: Received message: Message({Partition(topic=Topic(name='events'), index=0): 165})
21:36:47 post-process-forwarder-errors         | 21:36:47 [INFO] sentry.eventstream.kafka.dispatch: Received message: Message({Partition(topic=Topic(name='events'), index=0): 166})
21:36:47 post-process-forwarder-errors         | 21:36:47 [INFO] sentry.eventstream.kafka.dispatch: Received message: Message({Partition(topic=Topic(name='events'), index=0): 167})
21:36:50 post-process-forwarder-errors         | 21:36:50 [INFO] sentry.eventstream.kafka.dispatch: Received message: Message({Partition(topic=Topic(name='events'), index=0): 168})
21:36:50 post-process-forwarder-errors         | 21:36:50 [INFO] sentry.eventstream.kafka.dispatch: Received message: Message({Partition(topic=Topic(name='events'), index=0): 169})
21:36:50 post-process-forwarder-errors         | 21:36:50 [INFO] sentry.eventstream.kafka.dispatch: Received message: Message({Partition(topic=Topic(name='events'), index=0): 170})
21:36:52 post-process-forwarder-errors         | 21:36:52 [INFO] sentry.eventstream.kafka.dispatch: Received message: Message({Partition(topic=Topic(name='events'), index=0): 171})
21:36:52 post-process-forwarder-errors         | 21:36:52 [INFO] sentry.eventstream.kafka.dispatch: Received message: Message({Partition(topic=Topic(name='events'), index=0): 172})
21:36:54 post-process-forwarder-errors         | 21:36:54 [INFO] sentry.eventstream.kafka.dispatch: Received message: Message({Partition(topic=Topic(name='events'), index=0): 173})
21:36:57 post-process-forwarder-errors         | 21:36:57 [INFO] sentry.eventstream.kafka.dispatch: Received message: Message({Partition(topic=Topic(name='events'), index=0): 174})
21:36:57 post-process-forwarder-errors         | 21:36:57 [INFO] sentry.eventstream.kafka.dispatch: Received message: Message({Partition(topic=Topic(name='events'), index=0): 175})
21:36:59 post-process-forwarder-errors         | 21:36:59 [INFO] sentry.eventstream.kafka.dispatch: Received message: Message({Partition(topic=Topic(name='events'), index=0): 176})
21:36:59 post-process-forwarder-errors         | 21:36:59 [INFO] sentry.eventstream.kafka.dispatch: Received message: Message({Partition(topic=Topic(name='events'), index=0): 177})


With multiprocess strategy

21:27:32 system                                | post-process-forwarder-errors started (pid=45424)
21:27:48 post-process-forwarder-errors         | 21:27:48 [INFO] arroyo.processing.processor: New partitions assigned: {Partition(topic=Topic(name='events'), index=0): 134}
21:27:48 post-process-forwarder-errors         | 21:27:48 [INFO] sentry.post_process_forwarder.post_process_forwarder: Starting multiprocess post process forwarder
21:28:23 post-process-forwarder-errors         | 21:28:23 [INFO] sentry.eventstream.kafka.dispatch: Received message: Message({Partition(topic=Topic(name='events'), index=0): 135})
21:28:26 post-process-forwarder-errors         | 21:28:26 [INFO] sentry.eventstream.kafka.dispatch: Received message: Message({Partition(topic=Topic(name='events'), index=0): 136})
21:28:26 post-process-forwarder-errors         | 21:28:26 [INFO] sentry.eventstream.kafka.dispatch: Received message: Message({Partition(topic=Topic(name='events'), index=0): 137})
21:28:26 post-process-forwarder-errors         | 21:28:26 [INFO] sentry.eventstream.kafka.dispatch: Received message: Message({Partition(topic=Topic(name='events'), index=0): 138})
21:28:33 post-process-forwarder-errors         | 21:28:33 [INFO] sentry.eventstream.kafka.dispatch: Received message: Message({Partition(topic=Topic(name='events'), index=0): 139})
21:28:33 post-process-forwarder-errors         | 21:28:33 [INFO] sentry.eventstream.kafka.dispatch: Received message: Message({Partition(topic=Topic(name='events'), index=0): 140})
21:28:33 post-process-forwarder-errors         | 21:28:33 [INFO] sentry.eventstream.kafka.dispatch: Received message: Message({Partition(topic=Topic(name='events'), index=0): 141})
21:28:33 post-process-forwarder-errors         | 21:28:33 [INFO] sentry.eventstream.kafka.dispatch: Received message: Message({Partition(topic=Topic(name='events'), index=0): 142})
21:28:33 post-process-forwarder-errors         | 21:28:33 [INFO] sentry.eventstream.kafka.dispatch: Received message: Message({Partition(topic=Topic(name='events'), index=0): 143})
21:28:35 post-process-forwarder-errors         | 21:28:35 [INFO] sentry.eventstream.kafka.dispatch: Received message: Message({Partition(topic=Topic(name='events'), index=0): 144})
21:28:35 post-process-forwarder-errors         | 21:28:35 [INFO] sentry.eventstream.kafka.dispatch: Received message: Message({Partition(topic=Topic(name='events'), index=0): 145})
21:28:35 post-process-forwarder-errors         | 21:28:35 [INFO] sentry.eventstream.kafka.dispatch: Received message: Message({Partition(topic=Topic(name='events'), index=0): 146})
21:28:35 post-process-forwarder-errors         | 21:28:35 [INFO] sentry.eventstream.kafka.dispatch: Received message: Message({Partition(topic=Topic(name='events'), index=0): 147})
21:28:37 post-process-forwarder-errors         | 21:28:37 [INFO] sentry.eventstream.kafka.dispatch: Received message: Message({Partition(topic=Topic(name='events'), index=0): 148})
21:28:37 post-process-forwarder-errors         | 21:28:37 [INFO] sentry.eventstream.kafka.dispatch: Received message: Message({Partition(topic=Topic(name='events'), index=0): 149})
21:28:37 post-process-forwarder-errors         | 21:28:37 [INFO] sentry.eventstream.kafka.dispatch: Received message: Message({Partition(topic=Topic(name='events'), index=0): 150})
21:28:38 system                                | sending SIGTERM to post-process-forwarder-errors (pid 45424)
21:28:38 post-process-forwarder-errors         | 21:28:38 [INFO] arroyo.processing.processor: Shutdown signalled
21:28:38 post-process-forwarder-errors         | Traceback (most recent call last):
21:28:38 post-process-forwarder-errors         |   File "/Users/nikharsaxena/workspace/sentry/.venv/lib/python3.8/site-packages/arroyo/processing/processor.py", line 294, in run
21:28:38 post-process-forwarder-errors         |     self._run_once()
21:28:38 post-process-forwarder-errors         |   File "/Users/nikharsaxena/workspace/sentry/.venv/lib/python3.8/site-packages/arroyo/processing/processor.py", line 356, in _run_once
21:28:38 post-process-forwarder-errors         |     self.__message = self.__consumer.poll(timeout=1.0)
21:28:38 post-process-forwarder-errors         |   File "/Users/nikharsaxena/workspace/sentry/src/sentry/consumers/synchronized.py", line 247, in poll
21:28:38 post-process-forwarder-errors         |     message = self.__consumer.poll(timeout)
21:28:38 post-process-forwarder-errors         |   File "/Users/nikharsaxena/workspace/sentry/.venv/lib/python3.8/site-packages/arroyo/backends/kafka/consumer.py", line 394, in poll
21:28:38 post-process-forwarder-errors         |     message: Optional[ConfluentMessage] = self.__consumer.poll(
21:28:38 post-process-forwarder-errors         |   File "/Users/nikharsaxena/workspace/sentry/.venv/lib/python3.8/site-packages/arroyo/backends/kafka/configuration.py", line 45, in stats_callback
21:28:38 post-process-forwarder-errors         |     def stats_callback(stats_json: str) -> None:
21:28:38 post-process-forwarder-errors         |   File "/Users/nikharsaxena/workspace/sentry/.venv/lib/python3.8/site-packages/arroyo/processing/strategies/run_task_with_multiprocessing.py", line 509, in handle_sigchld
21:28:38 post-process-forwarder-errors         |     raise ChildProcessTerminated(signum)
21:28:38 post-process-forwarder-errors         | arroyo.processing.strategies.run_task_with_multiprocessing.ChildProcessTerminated: 20
21:28:38 post-process-forwarder-errors         | 21:28:38 [ERROR] arroyo.processing.processor: Caught exception, shutting down...
21:28:38 post-process-forwarder-errors         | 21:28:38 [INFO] arroyo.processing.strategies.run_task_with_multiprocessing: Terminating <multiprocessing.pool.Pool state=RUN pool_size=1>...
21:28:44 system                                | sending SIGKILL to post-process-forwarder-errors (pid 45424)
21:28:44 system                                | sending SIGKILL to post-process-forwarder-errors (pid 45424)
21:28:44 system                                | sending SIGKILL to post-process-forwarder-errors (pid 45424)
21:28:44 system                                | sending SIGKILL to post-process-forwarder-errors (pid 45424)
21:28:44 system                                | sending SIGKILL to post-process-forwarder-errors (pid 45424)
21:28:44 system                                | sending SIGKILL to post-process-forwarder-errors (pid 45424)
21:28:44 system                                | sending SIGKILL to post-process-forwarder-errors (pid 45424)
21:28:44 system                                | sending SIGKILL to post-process-forwarder-errors (pid 45424)
21:28:44 system                                | post-process-forwarder-errors stopped (rc=-9)

```

### CODE REVIEW

**Review by evanh — APPROVED** [review]
Is it possible to add a test for this? Or maybe run some tests in multithreaded and multiprocess mode?

File: src/sentry/post_process_forwarder/post_process_forwarder.py [hunk 1]
```diff
@@ -142,18 +142,48 @@ class PostProcessForwarderStrategyFactory(ProcessingStrategyFactory[KafkaPayload
...
         raise NotImplementedError()
 
-    def __init__(self, concurrency: int):
-        self.__concurrency = concurrency
-        self.__max_pending_futures = concurrency + 1000
+    def __init__(
+        self,
+        mode: str,
+        num_processes: int,
+        input_block_size: int,
```
Comment (untitaker) [inline src/sentry/post_process_forwarder/post_process_forwarder.py:149]: all of those arguments are now optional to the strategy. i would recommend to leave them at defaults of `None`, and change the CLI (of all consumers) to accept that default

Comment (nikhars) [reply src/sentry/post_process_forwarder/post_process_forwarder.py:149]: Not sure what you meant here. The multiprocessing options [here](https://github.com/getsentry/sentry/blob/d87bc5a8bdde9feb2ec24e57838c99b9b2266f3b/src/sentry/consumers/__init__.py#L34-L54) passes the default values to the constructor. And the call made from the new unified constructor passes all the parameters to the constructor from [here](https://github.com/getsentry/sentry/blob/d87bc5a8bdde9feb2ec24e57838c99b9b2266f3b/src/sentry/consumers/__init__.py#L34-L54). Wouldn't the logic of resolving which click option to be passed already be handled by existing code? Or am I missing some context?

Comment (untitaker) [reply src/sentry/post_process_forwarder/post_process_forwarder.py:149]: I am following up on this in https://github.com/getsentry/sentry/pull/61054

**Comment by nikhars** [comment]
> Is it possible to add a test for this? Or maybe run some tests in multithreaded and multiprocess mode?

Once [this](https://github.com/getsentry/sentry/pull/61021) PR gets merged, I will add a test for the multiprocess mode in the existing test.

**Pull request merged**